### PR TITLE
Fix an issue with simultaneous playback of incoming messages from 

### PR DESCRIPTION
**Pull request by megamk**
Fix an issue with simultaneous playback of incoming messages from  different channels.
The problem was that `ondata` handler was set up by overriding prototype which is not correct. 
To make it work new option `noPersistentPlayer` will need to be set to `true`

### CODE REVIEW

File: docs/js/incomingMessage.js.html [hunk 1]
```diff
@@ -148,20 +149,23 @@ <h1 class="page-title">Source: incomingMessage.js</h1>
     if (!this.options.decoder) {
       return;
     }
-    this.options.decoder.prototype.ondata = (pcmData) => {
+    this.decoder = new this.options.decoder(this.options);
+    this.decoder.ondata = (pcmData) => {
+      console.log(Date.now(), this.streamId);
```
Comment (vocoded) [inline docs/js/incomingMessage.js.html:154]: assume we need to remove

File: docs/js/incomingMessage.js.html [hunk 2]
```diff
@@ -148,20 +149,23 @@ <h1 class="page-title">Source: incomingMessage.js</h1>
...
       /**
        * Incoming voice message packet decoded
        * @event IncomingMessage#incoming_voice_data_decoded
        * @param {Float32Array} pcmData decoded pcm packet
        */
-      this.emit(Constants.EVENT_INCOMING_VOICE_DATA_DECODED, pcmData);
-      this.session.onIncomingVoiceDecoded(pcmData, this);
-    };
-    this.decoder = new this.options.decoder(this.options);
+      setTimeout(() => {
```
Comment (vocoded) [inline docs/js/incomingMessage.js.html:160]: is the yield required?

File: sdks/js/src/classes/incomingMessage.js [hunk 3]
```diff
@@ -120,20 +121,23 @@ class IncomingMessage extends Emitter {
     if (!this.options.decoder) {
       return;
     }
-    this.options.decoder.prototype.ondata = (pcmData) => {
+    this.decoder = new this.options.decoder(this.options);
+    this.decoder.ondata = (pcmData) => {
+      console.log(Date.now(), this.streamId);
```
Comment (vocoded) [inline sdks/js/src/classes/incomingMessage.js:126]: remove?

File: sdks/js/src/classes/incomingMessage.js [hunk 4]
```diff
@@ -120,20 +121,23 @@ class IncomingMessage extends Emitter {
...
+      setTimeout(() => {
+        this.emit(Constants.EVENT_INCOMING_VOICE_DATA_DECODED, pcmData);
+        this.session.onIncomingVoiceDecoded(pcmData, this);
+      }, 0);
+    }
   }
 
   initPlayer(sampleRate) {
-    if (IncomingMessage.PersistentPlayer) {
+    if (IncomingMessage.PersistentPlayer && !this.options.noPersistentPlayer) {
```
Comment (vocoded) [inline sdks/js/src/classes/incomingMessage.js:140]: below we will be overriding `IncomingMessage.PersistentPlayer` if this flag is set, which seems unexpected even if it isn't used

File: sdks/js/src/classes/incomingMessage.js [hunk 5]
```diff
@@ -120,20 +121,23 @@ class IncomingMessage extends Emitter {
...
       /**
        * Incoming voice message packet decoded
        * @event IncomingMessage#incoming_voice_data_decoded
        * @param {Float32Array} pcmData decoded pcm packet
        */
-      this.emit(Constants.EVENT_INCOMING_VOICE_DATA_DECODED, pcmData);
-      this.session.onIncomingVoiceDecoded(pcmData, this);
-    };
-    this.decoder = new this.options.decoder(this.options);
+      setTimeout(() => {
```
Comment (vocoded) [inline sdks/js/src/classes/incomingMessage.js:132]: same question as above, do we need to yield? if so could we use `setImmediate`?

**Review by vocoded — COMMENTED** [review]
couple of small things, thanks for finding this!

File: sdks/js/src/classes/incomingMessage.js (hunk 4 above)
Comment (megamk) [reply sdks/js/src/classes/incomingMessage.js:137]: added early return if it's not needed

**Comment by megamk** [comment]
@vocoded thanks for the review, ready for another look

File: sdks/js/src/classes/incomingMessage.js [hunk 6]
```diff
@@ -142,6 +143,9 @@ class IncomingMessage extends Emitter {
       return;
     }
     this.player = new this.options.player(this.options);
```
Comment (vocoded) [inline sdks/js/src/classes/incomingMessage.js:145]: do we need to set the sample rate as above?

**Review by vocoded — APPROVED** [review]

File: sdks/js/src/classes/incomingMessage.js (hunk 6 above)
Comment (megamk) [reply sdks/js/src/classes/incomingMessage.js:145]: no, `sampleRate` is a part of `this.options` and is set in the constructor

Comment (vocoded) [reply sdks/js/src/classes/incomingMessage.js:145]: sampleRate is passed as an argument to the function - are you saying we should ignore it? or should we have it override options if it is set?

Comment (megamk) [reply sdks/js/src/classes/incomingMessage.js:145]: yes, and that function is setting `this.options.sampleRate`. There's default value for it, so it's always defined. When using persistent player to play different messages, we need to change it to match a new message. If `noPersistentPlayer` is set to `true`, we don't need to override it

Comment (vocoded) [reply sdks/js/src/classes/incomingMessage.js:145]: ok, i see. it's a little confusing at a glance. another option is to remove the argument and change the logic in the conditional like:

```
if (IncomingMessage.PersistentPlayer && !this.options.noPersistentPlayer) {
  this.player = IncomingMessage.PersistentPlayer;
  this.player.setSampleRate(this.options.sampleRate);
  return;
}
```

Comment (megamk) [reply sdks/js/src/classes/incomingMessage.js:145]: @vocoded applied this change

**Review by vocoded — APPROVED** [review]

**Pull request merged**